internal/middleware/compression.go

```diff
@@ -15,12 +15,60 @@
 package middleware
 
 import (
+	"bytes"
 	"compress/gzip"
 	"compress/zlib"
 	"io"
 	"net/http"
 )
 
+const (
+	headerContentEncoding = "Content-Encoding"
+	headerValueGZIP       = "gzip"
+)
+
+type CompressRoundTripper struct {
+	http.RoundTripper
+	gzipWriter *gzip.Writer
+}
+
+func NewCompressRoundTripper(rt http.RoundTripper) *CompressRoundTripper {
+	return &CompressRoundTripper{
+		RoundTripper: rt,
+		gzipWriter:   gzip.NewWriter(nil),
+	}
+}
+
+func (r *CompressRoundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
+
+	// Gzip the body.
+	buf := bytes.NewBuffer([]byte{})
+	r.gzipWriter.Reset(buf)
```

Comment:
There is a problem here. It is not guaranteed that multiple goroutines are not calling this at the same time. So probably allocate a new zipper or use a pool

Reply:
Let me double check. I believe this is one roundtripper per http client and you are not allowed concurrent calls on the same http client.

Reply:
That can only be possible if the http client serializes all the requests

Reply:
OK, I can't find anything in the docs, so I think you are right since it is not prohibited. Let me fix.

Reply:
Changed to use a new zipper on every request. 
Not sure about the performance though, I don't know how costly it is to recreate gzipWriter on every request. If it shows to be slow may need to add a pool of gzip writers to use instead. I am keeping simple for now, can be improved later.